src/storage/index/hash_index.cpp

```diff
@@ -363,6 +418,198 @@ void HashIndex<T>::copyEntryToSlot(slot_id_t slotId, const T& entry, uint8_t fin
+    // temporary vector buffers instead of slots for improved performance) and then merge each of
+    // those one at a time into the disk slots. That will keep the low memory requirements and still
+    // let us update each on-disk slot one at a time.
+
+    auto diskSlotIterator = pSlots->iter();
+    // TODO: Use a separate random access iterator and one that's sequential for adding new overflow
+    // slots All new slots will be sequential and benefit from caching, but for existing randomly
+    // accessed slots we just benefit from the interface. However, the two iterators would not be
+    // able to pin the same page simultaneously
+    auto diskOverflowSlotIterator = oSlots->iter();
+    for (uint64_t slotId = 0; slotId < bulkInsertLocalStorage.pSlots->size(); slotId++) {
+        auto localSlot =
+            typename HashIndexBuilder<T>::SlotIterator(slotId, &bulkInsertLocalStorage);
+        // If mask is empty, skip this slot
+        if (!localSlot.slot->header.validityMask) {
+            // There should be no entries in the overflow slots, as we never leave gaps in the
+            // builder
+            continue;
+        }
+        diskSlotIterator.seek(slotId);
+        slot_id_t diskEntryPos = 0u;
+        auto* diskSlot = &*diskSlotIterator;
+        // Merge slot from local storage to existing slot
+        do {
+            for (auto entryPos = 0u; entryPos < getSlotCapacity<T>(); entryPos++) {
+                if (localSlot.slot->header.isEntryValid(entryPos)) {
+                    // Find the next empty entry, or add a new slot if there are no more entries
+                    while (diskSlot->header.isEntryValid(diskEntryPos) ||
+                           diskEntryPos >= getSlotCapacity<T>()) {
+                        diskEntryPos++;
+                        if (diskEntryPos >= getSlotCapacity<T>()) {
+                            if (diskSlot->header.nextOvfSlotId == 0) {
+                                // If there are no more disk slots in this chain, we need to add one
+                                diskSlot->header.nextOvfSlotId = diskOverflowSlotIterator.size();
+                                // This may invalidate diskSlot
+                                diskOverflowSlotIterator.pushBack();
+                            } else {
+                                diskOverflowSlotIterator.seek(diskSlot->header.nextOvfSlotId);
+                            }
+                            diskSlot = &*diskOverflowSlotIterator;
+                            diskEntryPos = 0;
+                        }
+                    }
+                    KU_ASSERT(diskEntryPos < getSlotCapacity<T>());
+                    copyAndUpdateSlotHeader<const T&, true>(*diskSlot, diskEntryPos,
+                        localSlot.slot->entries[entryPos].key, UINT32_MAX,
+                        localSlot.slot->header.fingerprints[entryPos]);
+                    KU_ASSERT([&]() {
+                        auto hash = hashStored(
```

Comment:
I feel we should directly store the hash inside localSlot, instead of recomputing it. Can you give this a try as an optimization later?

Reply:
Yes, that was part of what I was planning when I get to re-writing and optimizing the `HashIndexBuilder`, particularly since we need it to calculate the new primary slots if they don't match.